### PR TITLE
fix(server): update denyoom flag for all commands

### LINKED ISSUE #1632: introduce memory guards for memory eating commands
The current state of things - commands like lpush, zadd etc can add data into existing items, indefinitely to Dragonfly,  easily bypassing maxmemory constraints.

I suggest introducing at least the simplest safe-guard at generic, DispatchCommand level - 
if we *already* are beyond maxmemory limit and we are about to execute the "memory eating" command - let's return with OOM error.

What's a "memory-eating" command? We do have `CO::WRITE` flag annotating the write commands but it's actually not good. 

There are commands like `FLUSHDB`, `DEL`, `...POP...` etc that are write commands but they do not add memory pressure in the long term. We have been also marking commands with `CO::DENYOOM` but have not been actually using this flag. Moreover, for some commands, I think it's been attributed incorrectly (aka, ZINCRBY). I suggest that the semantics of this flag will be - "commands that should not execute if we are in oom state (beyond maxmemory capacity).. Currently `used_mem_current` atomic tracks current memory usage but would prefer we somehow avoid reading atomics on a hotpath. I have an idea how to do it we can discuss it later.

### PR DESCRIPTION
**Pull request by adiholden**
fixes #1632 
DENYOOM command flag marks which commands will be denied when memory usage crosses the "red zone".

### CODE REVIEW

File: src/server/json_family.cc [hunk 1]
```diff
@@ -1796,19 +1796,17 @@ void JsonFamily::Register(CommandRegistry* registry) {
   *registry << CI{"JSON.STRLEN", CO::READONLY | CO::FAST, 3, 1, 1, 1}.HFUNC(StrLen);
   *registry << CI{"JSON.OBJLEN", CO::READONLY | CO::FAST, 3, 1, 1, 1}.HFUNC(ObjLen);
   *registry << CI{"JSON.ARRLEN", CO::READONLY | CO::FAST, 3, 1, 1, 1}.HFUNC(ArrLen);
-  *registry << CI{"JSON.TOGGLE", CO::WRITE | CO::DENYOOM | CO::FAST, 3, 1, 1, 1}.HFUNC(Toggle);
-  *registry << CI{"JSON.NUMINCRBY", CO::WRITE | CO::DENYOOM | CO::FAST, 4, 1, 1, 1}.HFUNC(
-      NumIncrBy);
-  *registry << CI{"JSON.NUMMULTBY", CO::WRITE | CO::DENYOOM | CO::FAST, 4, 1, 1, 1}.HFUNC(
-      NumMultBy);
```
Comment (chakaz) [inline src/server/json_family.cc:1803]: What's the policy by which we decide on `DENYOOM`? What is `SETBIT` allowed and `JSON.NUMINCRBY` not?

File: src/server/json_family.cc [hunk 2]
```diff
@@ -1796,19 +1796,17 @@ void JsonFamily::Register(CommandRegistry* registry) {
...
+  *registry << CI{"JSON.NUMMULTBY", CO::WRITE | CO::FAST, 4, 1, 1, 1}.HFUNC(NumMultBy);
   *registry << CI{"JSON.DEL", CO::WRITE, -2, 1, 1, 1}.HFUNC(Del);
   *registry << CI{"JSON.FORGET", CO::WRITE, -2, 1, 1, 1}.HFUNC(Del);  // An alias of JSON.DEL.
   *registry << CI{"JSON.OBJKEYS", CO::READONLY | CO::FAST, 3, 1, 1, 1}.HFUNC(ObjKeys);
   *registry << CI{"JSON.STRAPPEND", CO::WRITE | CO::DENYOOM | CO::FAST, -4, 1, 1, 1}.HFUNC(
       StrAppend);
-  *registry << CI{"JSON.CLEAR", CO::WRITE | CO::DENYOOM | CO::FAST, 3, 1, 1, 1}.HFUNC(Clear);
-  *registry << CI{"JSON.ARRPOP", CO::WRITE | CO::DENYOOM | CO::FAST, -3, 1, 1, 1}.HFUNC(ArrPop);
-  *registry << CI{"JSON.ARRTRIM", CO::WRITE | CO::DENYOOM | CO::FAST, 5, 1, 1, 1}.HFUNC(ArrTrim);
+  *registry << CI{"JSON.CLEAR", CO::WRITE | CO::FAST, 3, 1, 1, 1}.HFUNC(Clear);
```
Comment (chakaz) [inline src/server/json_family.cc:1807]: This should clear memory, right? So why not allow it?

File: src/server/list_family.cc [hunk 3]
```diff
@@ -1315,14 +1315,13 @@ void ListFamily::Register(CommandRegistry* registry) {
       << CI{"LLEN", CO::READONLY | CO::FAST, 2, 1, 1, 1}.HFUNC(LLen)
       << CI{"LPOS", CO::READONLY | CO::FAST, -3, 1, 1, 1}.HFUNC(LPos)
       << CI{"LINDEX", CO::READONLY, 3, 1, 1, 1}.HFUNC(LIndex)
-      << CI{"LINSERT", CO::WRITE, 5, 1, 1, 1}.HFUNC(LInsert)
+      << CI{"LINSERT", CO::WRITE | CO::DENYOOM, 5, 1, 1, 1}.HFUNC(LInsert)
```
Comment (chakaz) [inline src/server/list_family.cc:1318]: This can increase memory usage, right? Why allow it then?

File: src/server/json_family.cc (hunk 1 above)
Comment (romange) [reply src/server/json_family.cc:1803]: https://redis.io/commands/json.numincrby/ 

`json.NUMINCRBY` increments the already existing field. Besides temporary allocation there is no memory growth here.

File: src/server/json_family.cc (hunk 2 above)
Comment (romange) [reply src/server/json_family.cc:1807]: we do allow it, this is why we removed DENYOOM

**Review by romange — APPROVED** [review]
@adiholden  please explain in the commit that denyoom flag says which commands can not run in the oom state where flag memory usage crosses the "red zone".

File: src/server/list_family.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/list_family.cc:1318]: Reply to all comments above: Its flipped 😄 Deny -> deny

**Review by dranikpg — DISMISSED** [review]
Didn't found conflicting cases from those changed. If only we could somehow reliably DCHECK it... 🤔

**Review by chakaz — DISMISSED** [review]
:facepalm: 
Re/ the failure: it seems unrelated, I just sent out https://github.com/dragonflydb/dragonfly/pull/1653

**Comment by adiholden** [comment]
@dranikpg @chakaz can you please approve, unless you have any comment

**Review by chakaz — APPROVED** [review]

**Pull request merged**